### PR TITLE
[typing] Add type hints to `__init__` methods in `torch.distributions`.

### LINKED ISSUE #144196: [typing] Add static type hints to `torch.distributions`.
### 🚀 The feature, motivation and pitch

Current lack of type hints causes some issues, for instance #76772

- [x] Add type hints for `lazy_property` class (#144110)
- [x] Add type hints for `@property` and `@lazy_property` (#144110)
- [ ] Add type hints to `__init__` methods (#144197)
- [ ] Add type hints to methods (#144219)
- [ ] Add type hints to attributes (#144219)
- [ ] Add type hints to utility functions (#144219)

### Alternatives

_No response_

### Additional context

_No response_

cc @fritzo @neerajprad @alicanb @nikitaved @ezyang @malfet @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by randolf-scholz**
Fixes #144196
Extends #144106 and #144110

## Open Problems:

- [x] Annotating with `numbers.Number` is a bad idea, should consider using `float`, `SupportsFloat` or some `Procotol`. https://github.com/pytorch/pytorch/pull/144197#discussion_r1903324769

# Notes

- `beta.py`: needed to add `type: ignore` since `broadcast_all` is untyped.
- `categorical.py`: converted `else` branches of mutually exclusive arguments to `if` branch[^2].
- ~~`dirichlet.py`: replaced `axis` with `dim` arguments.~~ #144402
- `gemoetric.py`: converted `else` branches of mutually exclusive arguments to `if` branch[^2].
- ~~`independent.py`: fixed bug in `Independent.__init__` where `tuple[int, ...]` could be passed to `Distribution.__init__` instead of `torch.Size`.~~ **EDIT:** turns out the bug is related to typing of `torch.Size`. #144218
- `independent.py`: made `Independent` a generic class of its base distribution.
- `multivariate_normal.py`: converted `else` branches of mutually exclusive arguments to `if` branch[^2].
- `relaxed_bernoulli.py`: added class-level type hint for `base_dist`.
- `relaxed_categorical.py`: added class-level type hint for `base_dist`.
- ~~`transforms.py`: Added missing argument to docstring of `ReshapeTransform`~~ #144401
- ~~`transforms.py`: Fixed bug in `AffineTransform.sign` (could return `Tensor` instead of `int`).~~ #144400
- `transforms.py`: Added `type: ignore` comments to `AffineTransform.log_abs_det_jacobian`[^1]; replaced `torch.abs(scale)` with `scale.abs()`.
- `transforms.py`: Added `type: ignore` comments to `AffineTransform.__eq__`[^1].
- `transforms.py`: Fixed type hint on `CumulativeDistributionTransform.domain`. Note that this is still an LSP violation, because `Transform.domain` is defined as `Constraint`, but `Distribution.domain` is defined as `Optional[Constraint]`.
- skipped: `constraints.py`, `constraints_registry.py`, `kl.py`, `utils.py`, `exp_family.py`, `__init__.py`.

## Remark

`TransformedDistribution`: `__init__` uses the check `if reinterpreted_batch_ndims > 0:`, which can lead to the creation of `Independent` distributions with only 1 component. This results in awkward code like `base_dist.base_dist` in `LogisticNormal`.

```python
import torch
from torch.distributions import *
b1 = Normal(torch.tensor([0.0]), torch.tensor([1.0]))
b2 = MultivariateNormal(torch.tensor([0.0]), torch.eye(1))
t = StickBreakingTransform()
d1 = TransformedDistribution(b1, t)
d2 = TransformedDistribution(b2, t)
print(d1.base_dist)  # Independent with 1 dimension
print(d2.base_dist)  # MultivariateNormal
```

One could consider changing this to `if reinterpreted_batch_ndims > 1:`. 

[^1]: Usage of `isinstance(value, numbers.Real)` leads to problems with static typing, as the `numbers` module is not supported by `mypy` (see <https://github.com/python/mypy/issues/3186>). This results in us having to add type-ignore comments in several places
[^2]: Otherwise, we would have to add a bunch of `type: ignore` comments to make `mypy` happy, as it isn't able to perform the type narrowing. Ideally, such code should be replaced with structural pattern matching once support for Python 3.9 is dropped.

cc @fritzo @neerajprad @alicanb @nikitaved @ezyang @malfet @xuzhao9 @gramster

### CODE REVIEW

**Comment by cyyever** [comment]
Use ```pyupgrade --py39-plus xxxx.py``` for faster Python 3.9 refactoring?

**Review by Skylion007 — COMMENTED** [review]
nits

File: torch/distributions/wishart.py [hunk 1]
```diff
@@ -76,12 +76,12 @@ class Wishart(ExponentialFamily):
...
         self,
-        df: Union[torch.Tensor, Number],
-        covariance_matrix: Optional[torch.Tensor] = None,
-        precision_matrix: Optional[torch.Tensor] = None,
-        scale_tril: Optional[torch.Tensor] = None,
+        df: Union[Tensor, Number],
+        covariance_matrix: Optional[Tensor] = None,
+        precision_matrix: Optional[Tensor] = None,
+        scale_tril: Optional[Tensor] = None,
         validate_args=None,
```
[the remark this answers is not in the document]
Comment (randolf-scholz) [reply torch/distributions/wishart.py]: I actually just noticed that some minutes ago :)

File: torch/distributions/inverse_gamma.py [hunk 2]
```diff
@@ -1,5 +1,8 @@
 # mypy: allow-untyped-defs
```
[the remark this answers is not in the document]
Comment (randolf-scholz) [reply torch/distributions/inverse_gamma.py:1]: I have another branch where I am tracking full typing, it's close to done, but there are some issues left

File: torch/distributions/transforms.py [hunk 3]
```diff
@@ -413,7 +416,12 @@ class IndependentTransform(Transform):
             dimensions to treat as dependent.
     """
 
-    def __init__(self, base_transform, reinterpreted_batch_ndims, cache_size=0):
+    def __init__(
+        self,
+        base_transform: Transform,
+        reinterpreted_batch_ndims: int,
+        cache_size=0,
```
[the remark this answers is not in the document]
Comment (randolf-scholz) [reply torch/distributions/transforms.py]: done

File: torch/distributions/inverse_gamma.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (randolf-scholz) [reply torch/distributions/inverse_gamma.py:1]: #144219

File: torch/distributions/bernoulli.py [hunk 4]
```diff
@@ -41,7 +42,12 @@ class Bernoulli(ExponentialFamily):
     has_enumerate_support = True
     _mean_carrier_measure = 0
 
-    def __init__(self, probs=None, logits=None, validate_args=None):
+    def __init__(
+        self,
+        probs: Optional[Union[Tensor, Number]] = None,
```
Comment (randolf-scholz) [inline torch/distributions/bernoulli.py:47]: Actually, I think all of these should be replaced by `Union[Tensor, float]`, annotating as `Number` is aweful, because it is incompatible with `mypy`:
https://mypy-play.net/?mypy=latest&python=3.12&gist=87f5f362040b0d56bca3d3468d0224f9

Comment (randolf-scholz) [reply torch/distributions/bernoulli.py:47]: I used `Number` because that is what the docstring says. Does `torch` try to support `numpy` scalars, or why is `Number` used instead of simple `float`?

Comment (randolf-scholz) [reply torch/distributions/bernoulli.py:47]: @Skylion007 If one wants to be generic, one could use `typing.SupportsFloat` which would support 3rd party float types.

File: torch/distributions/transforms.py [hunk 5]
```diff
@@ -776,7 +797,7 @@ def __eq__(self, other):
     def sign(self) -> int:
         if isinstance(self.scale, numbers.Real):
             return 1 if float(self.scale) > 0 else -1 if float(self.scale) < 0 else 0
-        return self.scale.sign()
+        return int(self.scale.sign().prod().item())  # type: ignore[union-attr]
```
Comment (malfet) [inline torch/distributions/transforms.py:800]: Please move this one to a separate PR. It's a legitimate bugfix, but should not be part of type annotation changes

File: torch/distributions/multivariate_normal.py [hunk 6]
```diff
@@ -165,7 +166,7 @@ def __init__(
                 covariance_matrix.shape[:-2], loc.shape[:-1]
             )
             self.covariance_matrix = covariance_matrix.expand(batch_shape + (-1, -1))
-        else:
+        elif precision_matrix is not None:
```
Comment (malfet) [inline torch/distributions/multivariate_normal.py:169]: Legitimate fix, but should not be part of type annotation change

File: torch/distributions/transforms.py [hunk 7]
```diff
@@ -480,11 +488,18 @@ class ReshapeTransform(Transform):
     Arguments:
         in_shape (torch.Size): The input event shape.
         out_shape (torch.Size): The output event shape.
+        cache_size (int): Size of cache. If zero, no caching is done. If one,
+            the latest single value is cached. Only 0 and 1 are supported.
```
Comment (malfet) [inline torch/distributions/transforms.py:492]: Thank you for the doc fix, but please do it in separate PR

**Review by malfet — CHANGES_REQUESTED** [review]
Thank you for the PR, but please split this one in purely type annotation, doc fixes and bug fixes (when you add an extra `elif:else:` clause and when you fix the semantic of `sign` method)

File: torch/distributions/multivariate_normal.py (hunk 6 above)
Comment (randolf-scholz) [reply torch/distributions/multivariate_normal.py:169]: This change is only necessary to make `mypy` happy once the arguments of `__init__` are annotated, it is not a bug fix.

See the notes in my OP, especially footnote 1.

File: torch/distributions/transforms.py (hunk 5 above)
Comment (randolf-scholz) [reply torch/distributions/transforms.py:800]: #144400

File: torch/distributions/transforms.py (hunk 7 above)
Comment (randolf-scholz) [reply torch/distributions/transforms.py:492]: #144401

File: torch/distributions/transforms.py (hunk 5 above)
Comment (randolf-scholz) [reply torch/distributions/transforms.py:800]: Upon second look, and as indicated by the test results, it seems like the class is designed to operate in a multihead fashion by default, so the returned tensor is actually the sign of the determinant for each of the heads. But

1. This is still an LSP violation
2. This still gives incorrect result when `event_dim > 0` (`prod` should be applied over the last `event_dim` axes.)

**Comment by randolf-scholz** [comment]
@malfet @Skylion007 I opened #144788 to address the `numbers.Number` issue more generally. Would it be fine for the moment to change type annotation simply to `float`, even when the docstring actually says `Number`?

Regarding testing the annotations, generally one of the simpler ways is to simply type check the test files.

**Comment by malfet** [comment]
@randolf-scholz if you think this PR is ready for review, please resolve merge conflicts

**Comment by randolf-scholz** [comment]
@malfet It is not ready due to https://github.com/pytorch/pytorch/pull/144197#discussion_r1903324769

File: torch/distributions/bernoulli.py (hunk 4 above)
Comment (randolf-scholz) [reply torch/distributions/bernoulli.py:47]: See #144788

File: torch/distributions/categorical.py [hunk 8]
```diff
@@ -59,7 +66,7 @@ def __init__(self, probs=None, logits=None, validate_args=None):
             if probs.dim() < 1:
                 raise ValueError("`probs` parameter must be at least one-dimensional.")
             self.probs = probs / probs.sum(-1, keepdim=True)
-        else:
+        if logits is not None:  # Note: 'if is None' instead of 'else' to help mypy
```
Comment (malfet) [inline torch/distributions/categorical.py:69]: Isn't that a better way to satisfy linter?
```suggestion
        else:
            assert logits is not None
```

**Comment by malfet** [comment]
@randolf-scholz but I feel like 50+% of this change could be landed separately even with mypy suppression for the whole file, say something like torch/distributions/weibull.py which will make incremental reviews faster.

File: torch/distributions/categorical.py (hunk 8 above)
Comment (randolf-scholz) [reply torch/distributions/categorical.py]: AS per suggestion I changed to this style in several spots

**Comment by randolf-scholz** [comment]
I added annotations to `broadcast_all` in this PR, since this function is used in virtually all `__init__` methods, this required adding a few assert statements to make `mypy` happy when mutually exclusive arguments are used.

File: torch/distributions/utils.py [hunk 9]
```diff
@@ -23,7 +23,9 @@
 ]
 
 
-def broadcast_all(*values):
+# FIXME: Use (*values: *Ts) -> tuple[Tensor for T in Ts] if Mapping-Type is ever added.
+#   See https://github.com/python/typing/issues/1216#issuecomment-2126153831
+def broadcast_all(*values: Union[Tensor, Number]) -> tuple[Tensor, ...]:
```
Comment (randolf-scholz) [inline torch/distributions/utils.py:28]: Currently, the annotation does not cover the case of  `__torch_function__`.

This is somewhat outside the scope of this PR, I think possibly a `Protocol` type should be added to `torch.types` that covers the concrete semantics.

File: torch/distributions/bernoulli.py (hunk 4 above)
Comment (randolf-scholz) [reply torch/distributions/bernoulli.py:47]: I changed it to adding a single `assert precision_matrix is not None` as in the other thread.

File: torch/distributions/multivariate_normal.py (hunk 6 above)
Comment (randolf-scholz) [reply torch/distributions/multivariate_normal.py]: I changed it to adding a single `assert precision_matrix is not None` as in the other thread.

File: torch/distributions/bernoulli.py (hunk 4 above)
Comment (randolf-scholz) [reply torch/distributions/bernoulli.py:47]: Fixed by #145086

File: torch/distributions/transforms.py [hunk 10]
```diff
@@ -807,7 +807,7 @@ def log_abs_det_jacobian(self, x, y):
         if isinstance(scale, _Number):
             result = torch.full_like(x, math.log(abs(scale)))
         else:
-            result = scale.abs().log()
+            result = torch.abs(scale).log()
```
Comment (randolf-scholz) [inline torch/distributions/transforms.py:810]: No longer needed; thanks to #145674 mypy correctly infers Tensor in the else branch.

**Comment by Skylion007** [comment]
@randolf-scholz Mind doing a rebase?

**Review by malfet — APPROVED** [review]
